src/sinks/util/buffer/metrics.rs

```diff
@@ -0,0 +1,548 @@
+        let mut state = DefaultHasher::new();
+        self.hash(&mut state);
+        let hash1 = state.finish();
+
+        let mut state = DefaultHasher::new();
+        other.hash(&mut state);
+        let hash2 = state.finish();
+
+        hash1 == hash2
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct MetricBuffer {
+    state: IndexSet<MetricEntry>,
+    metrics: IndexSet<MetricEntry>,
+}
+
+impl MetricBuffer {
+    pub fn new() -> Self {
+        Self {
+            state: IndexSet::new(),
+            metrics: IndexSet::new(),
+        }
+    }
+}
+
+impl Batch for MetricBuffer {
+    type Input = Event;
+    type Output = Vec<Metric>;
+
+    fn len(&self) -> usize {
+        self.num_items()
+    }
+
+    fn push(&mut self, item: Self::Input) {
+        let item = item.into_metric();
+        let new = MetricEntry(item.clone());
+
+        match item {
+            // gauges are special because gauge values could come
+            // in deltas - relative increments or decrements
+            Metric::Gauge { .. } => {
+                if let Some(MetricEntry(mut existing)) = self.metrics.swap_take(&new) {
+                    existing.merge(&item);
+                    self.metrics.insert(MetricEntry(existing));
+                } else {
+                    // if the gauge is not present in active batch,
+                    // then we look it up in permanent state, where we keep track
+                    // of gauge values throughout the entire application uptime
+                    let mut initial = if let Some(default) = self.state.get(&new) {
+                        default.0.clone()
+                    } else {
+                        Metric::Gauge {
+                            name: String::from(""),
+                            val: 0.0,
+                            direction: None,
+                            timestamp: None,
+                            tags: None,
+                        }
+                    };
+                    initial.merge(&item);
+                    self.metrics.insert(MetricEntry(initial));
+                }
+            }
+            _ => {
+                if let Some(MetricEntry(mut existing)) = self.metrics.swap_take(&new) {
+                    existing.merge(&item);
+                    self.metrics.insert(MetricEntry(existing));
+                } else {
+                    self.metrics.insert(new);
+                }
+            }
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.num_items() == 0
+    }
+
+    fn fresh(&self) -> Self {
+        let mut state = self.state.clone();
+        for entry in self.metrics.iter() {
+            if entry.0.is_gauge() {
+                state.insert(entry.clone());
+            }
+        }
+
+        Self {
+            state,
+            metrics: IndexSet::new(),
+        }
+    }
+
+    fn finish(self) -> Self::Output {
+        self.metrics.into_iter().map(|e| e.0).collect()
```

Comment:
If we're worried about the `swap_take` and ordering, we could do a sort here. Not sure that it's worth it though.

Reply:
Yeah, I suppose we could use standard HashSet then?

Reply:
Yes, I think that's true. For some reason, I had thought that `swap_take` would maintain the order we wanted, but I don't think that's correct looking at it now.